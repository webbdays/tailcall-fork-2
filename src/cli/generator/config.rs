use std::collections::BTreeMap;
use std::env;
use std::marker::PhantomData;
use std::path::Path;

use derive_setters::Setters;
use path_clean::PathClean;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use url::Url;

use crate::core::config::transformer::Preset;
use crate::core::config::{self, ConfigReaderContext};
use crate::core::mustache::Mustache;
use crate::core::valid::{Valid, ValidateFrom, Validator};

#[derive(Deserialize, Serialize, Debug, Default, Setters)]
#[serde(rename_all = "camelCase")]
pub struct Config<Status = UnResolved> {
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub inputs: Vec<Input<Status>>,
    pub output: Output<Status>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preset: Option<PresetConfig>,
    pub schema: Schema,
}

#[derive(Clone, Deserialize, Serialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct PresetConfig {
    merge_type: Option<f32>,
    #[serde(rename = "consolidateURL")]
    consolidate_url: Option<f32>,
    use_better_names: Option<bool>,
    tree_shake: Option<bool>,
    unwrap_single_field_types: Option<bool>,
}

#[derive(Deserialize, Serialize, Debug, Default)]
#[serde(transparent)]
pub struct Location<A>(
    #[serde(skip_serializing_if = "Location::is_empty")] pub String,
    #[serde(skip)] PhantomData<A>,
);

#[derive(Deserialize, Serialize, Debug)]
#[serde(transparent)]
pub struct Headers<A>(
    #[serde(skip_serializing_if = "is_default")] pub Option<BTreeMap<String, String>>,
    #[serde(skip)] PhantomData<A>,
);

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Input<Status = UnResolved> {
    #[serde(flatten)]
    pub source: Source<Status>,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Source<Status = UnResolved> {
    #[serde(rename_all = "camelCase")]
    Curl {
        src: Location<Status>,
        headers: Headers<Status>,
        field_name: String,
    },
    Proto {
        src: Location<Status>,
    },
    Config {
        src: Location<Status>,
    },
}

#[derive(Deserialize, Serialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct Output<Status = UnResolved> {
    #[serde(skip_serializing_if = "Location::is_empty")]
    pub path: Location<Status>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<config::Source>,
}

#[derive(Debug)]
pub enum Resolved {}

#[derive(Serialize, Deserialize, Debug, JsonSchema, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct UnResolved {}

#[derive(Deserialize, Serialize, Debug, Default)]
pub struct Schema {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub query: Option<String>,
}

fn between(threshold: f32, min: f32, max: f32) -> Valid<(), String> {
    Valid::<(), String>::fail(format!(
        "Invalid threshold value ({:.2}). Allowed range is [{:.2} - {:.2}] inclusive.",
        threshold, min, max
    ))
    .when(|| !(min..=max).contains(&threshold))
}

impl ValidateFrom<PresetConfig> for Preset {
    type Error = String;
    fn validate_from(config: PresetConfig) -> Valid<Self, Self::Error> {
        let mut preset = Preset::new();

        if let Some(merge_type) = config.merge_type {
            preset = preset.merge_type(merge_type);
        }

        if let Some(consolidate_url) = config.consolidate_url {
            preset = preset.consolidate_url(consolidate_url);
        }

        if let Some(use_better_names) = config.use_better_names {
            preset = preset.use_better_names(use_better_names);
        }

        if let Some(unwrap_single_field_types) = config.unwrap_single_field_types {
            preset = preset.unwrap_single_field_types(unwrap_single_field_types);
        }

        if let Some(tree_shake) = config.tree_shake {
            preset = preset.tree_shake(tree_shake);
        }

        // TODO: The field names in trace should be inserted at compile time.
        Valid::succeed(preset)
            .and_then(|preset| {
                let merge_types_th = between(preset.merge_type, 0.0, 1.0).trace("mergeType");
                let consolidate_url_th =
                    between(preset.consolidate_url, 0.0, 1.0).trace("consolidateURL");

                merge_types_th.and(consolidate_url_th).map_to(preset)
            })
            .trace("preset")
    }
}

impl<A> Location<A> {
    fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl Location<UnResolved> {
    fn into_resolved(self, parent_dir: Option<&Path>) -> Location<Resolved> {
        let path = {
            let path = self.0.as_str();
            if Url::parse(path).is_ok() || Path::new(path).is_absolute() {
                path.to_string()
            } else {
                let parent_dir = parent_dir.unwrap_or(Path::new(""));
                let joined_path = parent_dir.join(path);
                if let Ok(abs_path) = std::fs::canonicalize(&joined_path) {
                    abs_path.to_string_lossy().to_string()
                } else if let Ok(cwd) = env::current_dir() {
                    cwd.join(joined_path).clean().to_string_lossy().to_string()
                } else {
                    joined_path.clean().to_string_lossy().to_string()
                }
            }
        };
        Location(path, PhantomData)
    }
}

impl<A> Headers<A> {
    pub fn headers(&self) -> &Option<BTreeMap<String, String>> {
        &self.0
    }
}

impl Headers<UnResolved> {
    pub fn resolve(
        self,
        reader_context: &ConfigReaderContext,
    ) -> anyhow::Result<Headers<Resolved>> {
        // Resolve the header values with mustache template.
        let resolved_headers = if let Some(headers_inner) = self.0 {
            let mut resolved_headers = BTreeMap::new();
            for (key, value) in headers_inner.into_iter() {
                let template = Mustache::parse(&value)?;
                let resolved_value = template.render(reader_context);
                resolved_headers.insert(key, resolved_value);
            }
            Some(resolved_headers)
        } else {
            None
        };

        Ok(Headers(resolved_headers, PhantomData))
    }
}

impl Output<UnResolved> {
    pub fn resolve(self, parent_dir: Option<&Path>) -> anyhow::Result<Output<Resolved>> {
        Ok(Output {
            format: self.format,
            path: self.path.into_resolved(parent_dir),
        })
    }
}

impl Source<UnResolved> {
    pub fn resolve(
        self,
        parent_dir: Option<&Path>,
        reader_context: &ConfigReaderContext,
    ) -> anyhow::Result<Source<Resolved>> {
        match self {
            Source::Curl { src, field_name, headers } => {
                let resolved_path = src.into_resolved(parent_dir);
                let resolved_headers = headers.resolve(reader_context)?;
                Ok(Source::Curl { src: resolved_path, field_name, headers: resolved_headers })
            }
            Source::Proto { src } => {
                let resolved_path = src.into_resolved(parent_dir);
                Ok(Source::Proto { src: resolved_path })
            }
            Source::Config { src } => {
                let resolved_path = src.into_resolved(parent_dir);
                Ok(Source::Config { src: resolved_path })
            }
        }
    }
}

impl Input<UnResolved> {
    pub fn resolve(
        self,
        parent_dir: Option<&Path>,
        reader_context: &ConfigReaderContext,
    ) -> anyhow::Result<Input<Resolved>> {
        let resolved_source = self.source.resolve(parent_dir, reader_context)?;
        Ok(Input { source: resolved_source })
    }
}

impl Config {
    /// Resolves all the relative paths present inside the GeneratorConfig.
    pub fn into_resolved(
        self,
        config_path: &str,
        reader_context: ConfigReaderContext,
    ) -> anyhow::Result<Config<Resolved>> {
        let parent_dir = Some(Path::new(config_path).parent().unwrap_or(Path::new("")));

        let inputs = self
            .inputs
            .into_iter()
            .map(|input| input.resolve(parent_dir, &reader_context))
            .collect::<anyhow::Result<Vec<Input<Resolved>>>>()?;

        let output = self.output.resolve(parent_dir)?;

        Ok(Config { inputs, output, schema: self.schema, preset: self.preset })
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;
    use std::sync::Arc;

    use pretty_assertions::assert_eq;

    use super::*;
    use crate::core::tests::TestEnvIO;
    use crate::core::valid::{ValidateInto, ValidationError, Validator};

    fn location<S: AsRef<str>>(s: S) -> Location<UnResolved> {
        Location(s.as_ref().to_string(), PhantomData)
    }

    fn to_headers(raw_headers: BTreeMap<String, String>) -> Headers<UnResolved> {
        Headers(Some(raw_headers), PhantomData)
    }

    #[test]
    fn test_headers_resolve() {
        let mut headers = BTreeMap::new();
        headers.insert(
            "Authorization".to_owned(),
            "Bearer {{.env.TOKEN}}".to_owned(),
        );

        let mut env_vars = HashMap::new();
        let token = "eyJhbGciOiJIUzI1NiIsInR5";
        env_vars.insert("TOKEN".to_owned(), token.to_owned());

        let unresolved_headers = to_headers(headers);

        let mut runtime = crate::core::runtime::test::init(None);
        runtime.env = Arc::new(TestEnvIO::init(env_vars));

        let reader_ctx = ConfigReaderContext {
            runtime: &runtime,
            vars: &Default::default(),
            headers: Default::default(),
        };

        let resolved_headers = unresolved_headers.resolve(&reader_ctx).unwrap();

        let expected = format!("Bearer {token}");
        let result = resolved_headers
            .headers()
            .to_owned()
            .unwrap()
            .get("Authorization")
            .unwrap()
            .to_owned();
        assert_eq!(
            result, expected,
            "Authorization header should be resolved correctly"
        );
    }

    #[test]
    fn test_config_codec() {
        let mut headers = BTreeMap::new();
        headers.insert("user-agent".to_owned(), "tailcall-v1".to_owned());
        let config = Config::default().inputs(vec![Input {
            source: Source::Curl {
                src: location("https://example.com"),
                headers: to_headers(headers),
                field_name: "test".to_string(),
            },
        }]);
        let actual = serde_json::to_string_pretty(&config).unwrap();
        insta::assert_snapshot!(actual)
    }

    #[test]
    fn should_fail_when_invalid_merge_type_threshold() {
        let config_preset = PresetConfig {
            tree_shake: None,
            use_better_names: None,
            merge_type: Some(2.0),
            consolidate_url: None,
            unwrap_single_field_types: None,
        };

        let transform_preset: Result<Preset, ValidationError<String>> =
            config_preset.validate_into().to_result();
        assert!(transform_preset.is_err());
    }

    #[test]
    fn should_use_user_provided_presets_when_provided() {
        let config_preset = PresetConfig {
            tree_shake: Some(true),
            use_better_names: Some(true),
            merge_type: Some(0.5),
            consolidate_url: Some(1.0),
            unwrap_single_field_types: None,
        };
        let transform_preset: Preset = config_preset.validate_into().to_result().unwrap();
        let expected_preset = Preset::new()
            .use_better_names(true)
            .tree_shake(true)
            .consolidate_url(1.0)
            .merge_type(0.5);
        assert_eq!(transform_preset, expected_preset);
    }

    #[test]
    fn test_location_resolve_with_url() {
        let json_source = r#""https://dummyjson.com/products""#;
        let de_source: Location<UnResolved> = serde_json::from_str(json_source).unwrap();
        let de_source = de_source.into_resolved(None);
        assert_eq!(de_source.0, "https://dummyjson.com/products");
        assert_eq!(de_source.1, PhantomData::<Resolved>);
    }

    #[test]
    fn test_is_empty() {
        let location_empty: Location<UnResolved> = serde_json::from_str(r#""""#).unwrap();
        let location_non_empty: Location<UnResolved> =
            serde_json::from_str(r#""https://dummyjson.com/products""#).unwrap();
        assert!(location_empty.is_empty());
        assert!(!location_non_empty.is_empty());
    }
}
