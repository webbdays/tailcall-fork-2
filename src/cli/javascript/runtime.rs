use std::cell::{OnceCell, RefCell};
use std::fmt::{Debug, Formatter};
use std::thread;

use async_graphql_value::ConstValue;
use rquickjs::{Context, Ctx, FromJs, Function, IntoJs, Value};

use crate::core::worker::{Command, Event, WorkerRequest};
use crate::core::{blueprint, worker, WorkerIO};

struct LocalRuntime(Context);

thread_local! {
    // Practically only one JS runtime is created for every Runtime because tokio_runtime is single threaded.
  static LOCAL_RUNTIME: RefCell<OnceCell<LocalRuntime>> = const { RefCell::new(OnceCell::new()) };
}

#[rquickjs::function]
fn qjs_print(msg: String, is_err: bool) {
    if is_err {
        tracing::error!("{msg}");
    } else {
        tracing::info!("{msg}");
    }
}

fn setup_builtins(ctx: &Ctx<'_>) -> rquickjs::Result<()> {
    ctx.globals().set("__qjs_print", js_qjs_print)?;
    let _: Value = ctx.eval_file("src/cli/javascript/shim/console.js")?;

    Ok(())
}

impl LocalRuntime {
    fn try_new(script: blueprint::Script) -> anyhow::Result<Self> {
        let source = script.source;
        let js_runtime = rquickjs::Runtime::new()?;
        let context = Context::full(&js_runtime)?;
        context.with(|ctx| {
            setup_builtins(&ctx)?;
            ctx.eval(source)
        })?;

        tracing::debug!("JS Runtime created: {:?}", thread::current().name());
        Ok(Self(context))
    }
}

pub struct Runtime {
    script: blueprint::Script,
    // Single threaded JS runtime, that's shared across all tokio workers.
    tokio_runtime: Option<tokio::runtime::Runtime>,
}

impl Debug for Runtime {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Runtime {{ script: {:?} }}", self.script)
    }
}

impl Runtime {
    pub fn new(script: blueprint::Script) -> Self {
        let tokio_runtime = tokio::runtime::Builder::new_multi_thread()
            .worker_threads(1)
            .build()
            .expect("JS runtime not initialized");

        Self { script, tokio_runtime: Some(tokio_runtime) }
    }
}

impl Drop for Runtime {
    fn drop(&mut self) {
        // implicit call implementation to shutdown the tokio runtime
        // without blocking. Otherwise it will panic on an attempt to
        // drop AppContext in async runtime (e.g. in tests at least)
        if let Some(runtime) = self.tokio_runtime.take() {
            runtime.shutdown_background();
        }
    }
}

#[async_trait::async_trait]
impl WorkerIO<Event, Command> for Runtime {
    async fn call(&self, name: &str, event: Event) -> Result<Option<Command>, worker::Error> {
        let script = self.script.clone();
        let name = name.to_string(); // TODO
        if let Some(runtime) = &self.tokio_runtime {
            runtime
                .spawn(async move {
                    init_rt(script)?;
                    call(name, event)
                })
                .await?
        } else {
            Err(worker::Error::JsRuntimeStopped)
        }
    }
}

#[async_trait::async_trait]
impl WorkerIO<ConstValue, ConstValue> for Runtime {
    async fn call(
        &self,
        name: &str,
        input: ConstValue,
    ) -> Result<Option<ConstValue>, worker::Error> {
        let script = self.script.clone();
        let name = name.to_string();
        let value = serde_json::to_string(&input)?;
        if let Some(runtime) = &self.tokio_runtime {
            runtime
                .spawn(async move {
                    init_rt(script)?;
                    execute_inner(name, value).map(Some)
                })
                .await?
        } else {
            Err(worker::Error::JsRuntimeStopped)
        }
    }
}

fn init_rt(script: blueprint::Script) -> anyhow::Result<()> {
    // initialize runtime if this is the first call
    // exit if failed to initialize
    LOCAL_RUNTIME.with(move |cell| {
        if cell.borrow().get().is_none() {
            LocalRuntime::try_new(script).and_then(|runtime| {
                cell.borrow().set(runtime).map_err(|_| {
                    anyhow::anyhow!("trying to reinitialize an already initialized QuickJS runtime")
                })
            })
        } else {
            Ok(())
        }
    })
}

fn prepare_args<'js>(ctx: &Ctx<'js>, req: WorkerRequest) -> rquickjs::Result<(Value<'js>,)> {
    let object = rquickjs::Object::new(ctx.clone())?;
    object.set("request", req.into_js(ctx)?)?;
    Ok((object.into_value(),))
}

fn call(name: String, event: Event) -> Result<Option<Command>, worker::Error> {
    LOCAL_RUNTIME.with_borrow_mut(|cell| {
        let runtime = cell.get_mut().ok_or(worker::Error::RuntimeNotInitialized)?;
        runtime.0.with(|ctx| match event {
            Event::Request(req) => {
                let fn_as_value = ctx
                    .globals()
                    .get::<&str, Function>(name.as_str())
                    .map_err(|e| worker::Error::GlobalThisNotInitialised(e.to_string()))?;

                let function = fn_as_value
                    .as_function()
                    .ok_or(worker::Error::InvalidFunction(name))?;

                let args =
                    prepare_args(&ctx, req).map_err(|e| worker::Error::Rquickjs(e.to_string()))?;
                let command: Option<Value> = function.call(args).ok();
                command
                    .map(|output| Command::from_js(&ctx, output))
                    .transpose()
                    .map_err(|e| worker::Error::DeserializeFailed(e.to_string()))
            }
        })
    })
}

fn execute_inner(name: String, value: String) -> Result<ConstValue, worker::Error> {
    LOCAL_RUNTIME.with_borrow_mut(|cell| {
        let runtime = cell.get_mut().ok_or(worker::Error::RuntimeNotInitialized)?;
        runtime.0.with(|ctx| {
            let fn_as_value = ctx
                .globals()
                .get::<_, rquickjs::Function>(&name)
                .map_err(|e| worker::Error::GlobalThisNotInitialised(e.to_string()))?;

            let function = fn_as_value
                .as_function()
                .ok_or(worker::Error::InvalidFunction(name.clone()))?;
            let val: String = function
                .call((value,))
                .map_err(|e| worker::Error::FunctionValueParseError(e.to_string(), name))?;
            Ok::<_, worker::Error>(serde_json::from_str(&val)?)
        })
    })
}
