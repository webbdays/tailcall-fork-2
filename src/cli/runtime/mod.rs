mod env;
mod file;
mod http;

use std::fs;
use std::hash::Hash;
use std::sync::Arc;

pub use http::NativeHttp;
use inquire::{Confirm, Select};

use crate::core::blueprint::Blueprint;
use crate::core::cache::InMemoryCache;
use crate::core::runtime::TargetRuntime;
use crate::core::worker::{Command, Event};
use crate::core::{blueprint, EnvIO, FileIO, HttpIO, WorkerIO};

// Provides access to env in native rust environment
fn init_env() -> Arc<dyn EnvIO> {
    Arc::new(env::EnvNative::init())
}

// Provides access to file system in native rust environment
fn init_file() -> Arc<dyn FileIO> {
    Arc::new(file::NativeFileIO::init())
}

fn init_http_worker_io(
    script: Option<blueprint::Script>,
) -> Option<Arc<dyn WorkerIO<Event, Command>>> {
    #[cfg(feature = "js")]
    return Some(super::javascript::init_worker_io(script?));
    #[cfg(not(feature = "js"))]
    {
        let _ = script;
        None
    }
}

fn init_resolver_worker_io(
    script: Option<blueprint::Script>,
) -> Option<Arc<dyn WorkerIO<async_graphql::Value, async_graphql::Value>>> {
    #[cfg(feature = "js")]
    return Some(super::javascript::init_worker_io(script?));
    #[cfg(not(feature = "js"))]
    {
        let _ = script;
        None
    }
}

// Provides access to http in native rust environment
fn init_http(blueprint: &Blueprint) -> Arc<dyn HttpIO> {
    Arc::new(http::NativeHttp::init(
        &blueprint.upstream,
        &blueprint.telemetry,
    ))
}

// Provides access to http in native rust environment
fn init_http2_only(blueprint: &Blueprint) -> Arc<dyn HttpIO> {
    Arc::new(http::NativeHttp::init(
        &blueprint.upstream.clone().http2_only(true),
        &blueprint.telemetry,
    ))
}

fn init_in_memory_cache<K: Hash + Eq, V: Clone>() -> InMemoryCache<K, V> {
    InMemoryCache::new()
}

pub fn init(blueprint: &Blueprint) -> TargetRuntime {
    #[cfg(not(feature = "js"))]
    tracing::warn!("JS capabilities are disabled in this build");

    TargetRuntime {
        http: init_http(blueprint),
        http2_only: init_http2_only(blueprint),
        env: init_env(),
        file: init_file(),
        cache: Arc::new(init_in_memory_cache()),
        extensions: Arc::new(vec![]),
        cmd_worker: init_http_worker_io(blueprint.server.script.clone()),
        worker: init_resolver_worker_io(blueprint.server.script.clone()),
    }
}

pub async fn confirm_and_write(
    runtime: TargetRuntime,
    path: &str,
    content: &[u8],
) -> anyhow::Result<()> {
    let file_exists = fs::metadata(path).is_ok();

    if file_exists {
        let confirm = Confirm::new(&format!("Do you want to overwrite the file {path}?"))
            .with_default(false)
            .prompt()?;

        if !confirm {
            return Ok(());
        }
    }

    runtime.file.write(path, content).await?;

    Ok(())
}

pub async fn create_directory(folder_path: &str) -> anyhow::Result<()> {
    let folder_exists = fs::metadata(folder_path).is_ok();

    if !folder_exists {
        let confirm = Confirm::new(&format!(
            "Do you want to create the folder {}?",
            folder_path
        ))
        .with_default(false)
        .prompt()?;

        if confirm {
            fs::create_dir_all(folder_path)?;
        } else {
            return Ok(());
        };
    }

    Ok(())
}

pub fn select_prompt<T: std::fmt::Display>(message: &str, options: Vec<T>) -> anyhow::Result<T> {
    Ok(Select::new(message, options).prompt()?)
}
