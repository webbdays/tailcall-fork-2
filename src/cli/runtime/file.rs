use tokio::io::{AsyncReadExt, AsyncWriteExt};

use crate::cli::CLIError;
use crate::core::FileIO;

#[derive(Clone)]
pub struct NativeFileIO {}

impl NativeFileIO {
    pub fn init() -> Self {
        NativeFileIO {}
    }
}

async fn read(path: &str) -> anyhow::Result<String> {
    let mut file = tokio::fs::File::open(path).await?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).await?;
    Ok(String::from_utf8(buffer)?)
}

async fn write<'a>(path: &'a str, content: &'a [u8]) -> anyhow::Result<()> {
    let mut file = tokio::fs::File::create(path).await?;
    file.write_all(content).await?;
    Ok(())
}

#[async_trait::async_trait]
impl FileIO for NativeFileIO {
    async fn write<'a>(&'a self, path: &'a str, content: &'a [u8]) -> anyhow::Result<()> {
        write(path, content).await.map_err(|err| {
            CLIError::new(format!("Failed to write file: {}", path).as_str())
                .description(err.to_string())
        })?;
        tracing::info!("File write: {} ... ok", path);
        Ok(())
    }

    async fn read<'a>(&'a self, path: &'a str) -> anyhow::Result<String> {
        let content = read(path).await.map_err(|err| {
            CLIError::new(format!("Failed to read file: {}", path).as_str())
                .description(err.to_string())
        })?;
        tracing::info!("File read: {} ... ok", path);
        Ok(content)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_write_error() {
        // Attempt to write to an invalid path
        let file_io = NativeFileIO::init();
        let result = file_io.write("/invalid/path/to/file.txt", b"content").await;

        // Verify that an error is returned
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_read_error() {
        // Attempt to read from a non-existent file
        let file_io = NativeFileIO::init();
        let result = file_io.read("/non/existent/file.txt").await;

        // Verify that an error is returned
        assert!(result.is_err());
    }
}
