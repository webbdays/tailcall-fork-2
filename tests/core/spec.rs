extern crate core;

use std::collections::BTreeMap;
use std::path::Path;
use std::sync::Arc;
use std::{fs, panic};

use anyhow::Context;
use colored::Colorize;
use futures_util::future::join_all;
use hyper::{Body, Request};
use serde::{Deserialize, Serialize};
use tailcall::core::app_context::AppContext;
use tailcall::core::async_graphql_hyper::{GraphQLBatchRequest, GraphQLRequest};
use tailcall::core::blueprint::Blueprint;
use tailcall::core::config::reader::ConfigReader;
use tailcall::core::config::transformer::Required;
use tailcall::core::config::{Config, ConfigModule, Source};
use tailcall::core::http::handle_request;
use tailcall::core::merge_right::MergeRight;
use tailcall::core::print_schema::print_schema;
use tailcall::core::valid::{Cause, ValidationError, Validator};
use tailcall_prettier::Parser;

use super::file::File;
use super::http::Http;
use super::model::*;
use super::runtime::ExecutionSpec;
use crate::core::runtime;

#[derive(Debug, Default, Deserialize, Serialize, PartialEq)]
struct SDLError {
    message: String,
    trace: Vec<String>,
    description: Option<String>,
}

impl<'a> From<Cause<&'a str>> for SDLError {
    fn from(value: Cause<&'a str>) -> Self {
        SDLError {
            message: value.message.to_string(),
            trace: value.trace.iter().map(|e| e.to_string()).collect(),
            description: None,
        }
    }
}

impl From<Cause<String>> for SDLError {
    fn from(value: Cause<String>) -> Self {
        SDLError {
            message: value.message.to_string(),
            trace: value.trace.iter().map(|e| e.to_string()).collect(),
            description: value.description,
        }
    }
}

async fn is_sdl_error(spec: ExecutionSpec, mock_http_client: Arc<Http>) -> bool {
    if spec.sdl_error {
        // errors: errors are expected, make sure they match
        let (source, content) = &spec.server[0];

        let config = source.decode(content);

        let config = match config {
            Ok(config) => {
                let mut runtime = runtime::create_runtime(mock_http_client, spec.env.clone(), None);
                runtime.file = Arc::new(File::new(spec.clone()));
                let reader = ConfigReader::init(runtime);
                match reader.resolve(config, spec.path.parent()).await {
                    Ok(config) => Blueprint::try_from(&config),
                    Err(e) => Err(ValidationError::new(e.to_string())),
                }
            }
            Err(e) => Err(e),
        };

        match config {
            Ok(_) => {
                tracing::error!("\terror FAIL");
                panic!(
                    "Spec {} {:?} with \"sdl error\" directive did not have a validation error.",
                    spec.name, spec.path
                );
            }
            Err(cause) => {
                let errors: Vec<SDLError> =
                    cause.as_vec().iter().map(|e| e.to_owned().into()).collect();

                let snapshot_name = format!("{}_error", spec.safe_name);

                insta::assert_json_snapshot!(snapshot_name, errors);
            }
        };

        return true;
    }
    false
}

async fn check_server_config(spec: ExecutionSpec) -> Vec<Config> {
    let mut server: Vec<Config> = Vec::with_capacity(spec.server.len());

    for (i, (source, content)) in spec.server.iter().enumerate() {
        let config = Config::from_source(source.to_owned(), content).unwrap_or_else(|e| {
            panic!(
                "Couldn't parse GraphQL in server definition #{} of {:#?}: {}",
                i + 1,
                spec.path,
                e
            )
        });

        let config = Config::default().merge_right(config);

        // TODO: we should probably figure out a way to do this for every test
        // but GraphQL identity checking is very hard, since a lot depends on the code
        // style the re-serializing check gives us some of the advantages of the
        // identity check too, but we are missing out on some by having it only
        // enabled for either new tests that request it or old graphql_spec
        // tests that were explicitly written with it in mind
        if spec.check_identity {
            if matches!(source, Source::GraphQL) {
                let actual = config.to_sdl();

                // \r is added automatically in windows, it's safe to replace it with \n
                let content = content.replace("\r\n", "\n");

                let path_str = spec.path.display().to_string();
                let context = format!("path: {}", path_str);

                let actual = tailcall_prettier::format(actual, &tailcall_prettier::Parser::Gql)
                    .await
                    .map_err(|e| e.with_context(context.clone()))
                    .unwrap();

                let expected = tailcall_prettier::format(content, &tailcall_prettier::Parser::Gql)
                    .await
                    .map_err(|e| e.with_context(context.clone()))
                    .unwrap();

                pretty_assertions::assert_eq!(
                    actual,
                    expected,
                    "Identity check failed for {:#?}",
                    spec.path,
                );
            } else {
                panic!(
                    "Spec {:#?} has \"check identity\" enabled, but its config isn't in GraphQL.",
                    spec.path
                );
            }
        }

        server.push(config);
    }
    server
}

async fn run_query_tests_on_spec(
    spec: ExecutionSpec,
    server: Vec<ConfigModule>,
    mock_http_client: Arc<Http>,
) {
    if let Some(tests) = spec.test.as_ref() {
        let app_ctx = spec
            .app_context(
                server.first().unwrap(),
                spec.env.clone().unwrap_or_default(),
                mock_http_client.clone(),
            )
            .await;

        // test: Run test specs

        for (i, test) in tests.iter().enumerate() {
            let response = run_test(app_ctx.clone(), test)
                .await
                .context(spec.path.to_str().unwrap().to_string())
                .unwrap();

            let mut headers: BTreeMap<String, String> = BTreeMap::new();

            for (key, value) in response.headers() {
                headers.insert(key.to_string(), value.to_str().unwrap().to_string());
            }

            let response: APIResponse = APIResponse {
                status: response.status().clone().as_u16(),
                headers,
                body: Some(APIBody::Value(
                    serde_json::from_slice(
                        &hyper::body::to_bytes(response.into_body()).await.unwrap(),
                    )
                    .unwrap_or_default(),
                )),
            };

            let snapshot_name = format!("{}_{}", spec.safe_name, i);

            insta::assert_json_snapshot!(snapshot_name, response);
        }

        mock_http_client.test_hits(&spec.path);
    }
}

async fn test_spec(spec: ExecutionSpec) {
    // insta settings
    let mut insta_settings = insta::Settings::clone_current();
    insta_settings.set_prepend_module_to_snapshot(false);
    let _guard = insta::Settings::bind_to_scope(&insta_settings);

    let mock_http_client = Arc::new(Http::new(&spec));

    // check sdl error if any
    if is_sdl_error(spec.clone(), mock_http_client.clone()).await {
        return;
    }

    // Parse and validate all server configs + check for identity
    let server = check_server_config(spec.clone()).await;

    // Resolve all configs
    let mut runtime = runtime::create_runtime(mock_http_client.clone(), spec.env.clone(), None);
    runtime.file = Arc::new(File::new(spec.clone()));
    let reader = ConfigReader::init(runtime);

    let server: Vec<ConfigModule> = join_all(
        server
            .into_iter()
            .map(|config| reader.resolve(config, spec.path.parent())),
    )
    .await
    .into_iter()
    .enumerate()
    .map(|(i, result)| {
        result.unwrap_or_else(|e| {
            panic!(
                "Couldn't resolve GraphQL in server definition #{} of {:#?}: {}",
                i + 1,
                spec.path,
                e
            )
        })
    })
    .collect();

    // merged: Run merged specs
    let merged = server
        .iter()
        .fold(ConfigModule::default(), |acc, c| acc.merge_right(c.clone()))
        // Apply required transformers to the configuration
        .transform(Required)
        .to_result()
        .unwrap()
        .to_sdl();

    let formatter = tailcall_prettier::format(merged, &Parser::Gql)
        .await
        .unwrap();

    let snapshot_name = format!("{}_merged", spec.safe_name);

    insta::assert_snapshot!(snapshot_name, formatter);

    // client: Check if client spec matches snapshot
    if server.len() == 1 {
        let config = &server[0];

        let client = print_schema(
            (Blueprint::try_from(config)
                .context(format!("file: {}", spec.path.to_str().unwrap()))
                .unwrap())
            .to_schema(),
        );

        let formatted = tailcall_prettier::format(client, &Parser::Gql)
            .await
            .unwrap();
        let snapshot_name = format!("{}_client", spec.safe_name);

        insta::assert_snapshot!(snapshot_name, formatted);
    }

    // run query tests
    run_query_tests_on_spec(spec, server, mock_http_client).await;
}

pub async fn load_and_test_execution_spec(path: &Path) -> anyhow::Result<()> {
    let contents = fs::read_to_string(path)?;
    let spec: ExecutionSpec = ExecutionSpec::from_source(path, contents)
        .await
        .with_context(|| path.display().to_string())?;

    match spec.runner {
        Some(Annotation::Skip) => {
            println!("{} ... {}", spec.path.display(), "skipped".blue());
        }
        None => test_spec(spec).await,
    }

    Ok(())
}

async fn run_test(
    app_ctx: Arc<AppContext>,
    request: &APIRequest,
) -> anyhow::Result<hyper::Response<Body>> {
    let body = request
        .body
        .as_ref()
        .map(|body| Body::from(body.to_bytes()))
        .unwrap_or_default();

    let method = request.method.clone();
    let headers = request.headers.clone();
    let url = request.url.clone();
    let req = headers
        .into_iter()
        .fold(
            Request::builder()
                .method(method.to_hyper())
                .uri(url.as_str()),
            |acc, (key, value)| acc.header(key, value),
        )
        .body(body)?;

    // TODO: reuse logic from server.rs to select the correct handler
    if app_ctx.blueprint.server.enable_batch_requests {
        handle_request::<GraphQLBatchRequest>(req, app_ctx).await
    } else {
        handle_request::<GraphQLRequest>(req, app_ctx).await
    }
}
