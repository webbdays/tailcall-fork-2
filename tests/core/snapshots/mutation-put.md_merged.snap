---
source: tests/core/spec.rs
expression: formatter
---
schema @server @upstream(baseURL: "http://jsonplaceholder.typicode.com") {
  query: Query
  mutation: Mutation
}

input PostInput {
  body: String
  id: Int
  title: String
  userId: Int
}

type Mutation {
  insertPost(input: PostInput!): Post @http(body: "{{.args.input}}", method: "PUT", path: "/posts/{{.args.input.id}}")
}

type Post {
  body: String
  id: Int
  title: String
  userId: Int
}

type Query {
  firstUser: User @http(path: "/users/1")
}

type User {
  id: Int
  name: String
}
